API: Fix redirect issues

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

The 6.6.1 has the following issues:
- In [this](https://github.com/grafana/grafana/pull/21057) modification I had made the wrong assumption that the appSubUrl does not include the leading slash. As a result the redirect validation is not correct. In addition to this, validation should not consider correct paths like `/<appSubUrl>blahblah`.
- Frontend prepends the subpath to redirects in addition to the backend (`/grafana` ends up to `/grafana/grafana`)

[This](https://github.com/grafana/grafana/pull/21652) fix addresses the above issues by [removing](https://github.com/grafana/grafana/pull/21652/files#diff-c6731ee8c1631796c076df9ae9e4d2c7R50) the subpath from redirects in the backend.
However it does not work optimally when redirects occur in the backend before reaching the frontend ([like in the OAuth flow](https://github.com/grafana/grafana/issues/22227)) because the session cookie path is set to `appSubUrl/` and requests without the appSubUrl can not read it.

[This](https://github.com/grafana/grafana/pull/22265/) fixes the above issue by removing the trailing slash from the cookie path.

The reason for this is that addresses the causes of the problem instead of the symptoms.

This PR:
- [X] Reverts [this](https://github.com/grafana/grafana/pull/22265/) (#22671)
- [X] Fixes redirect validation and tests (#22675)
- [X] Modifies the backend and the frontend to prepend the appSubUrl only if it’s missing from the redirect URL (#22676)
- [X] Validates the redirect URL in OAuth login (#22677)
- [X] Fixes login view if `redirect_to` value is invalid but the user is authenticated so as the user is navigated to the home page instead of rendering login page with error. (#22678)

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #22227
Fixes #22461

**Special notes for your reviewer**:

I have tested it with  the following scenarios:
- with subpath `/grafana`:
    - base auth:
        - [X] Dashboard URL (http://localhost:3000/grafana/d/dFcV-J4Zz/16707?orgId=1)
        - [X] http://localhost:3000/grafana/
        - [X] http://localhost:3000/grafana
        - [X] http://localhost:3000/grafanagrafana
    - oauth:
        - [X] Dashboard URL (http://localhost:3000/grafana/d/dFcV-J4Zz/16707?orgId=1)
        - [X] http://localhost:3000/grafana/
        - [X] http://localhost:3000/grafana
        - [X] http://localhost:3000/grafanagrafana

- without subpath:
    - base auth:
        - [X]  Dashboard URL (http://localhost:3000/d/dFcV-J4Zz/16707?orgId=1)
        - [X] http://localhost:3000/
        - [X] http://localhost:3000
        - [X] http://localhost:3000/grafanagrafana
    - oauth:
        - [X]  Dashboard URL (http://localhost:3000/d/dFcV-J4Zz/16707?orgId=1)
        - [X] http://localhost:3000/
        - [X] http://localhost:3000
        - [X] http://localhost:3000/grafanagrafana

kylebrandt commented on pkg/api/login.go:
```diff
@@ -35,9 +35,7 @@ func (hs *HTTPServer) validateRedirectTo(redirectTo string) error {
-	// when using a subUrl, the redirect_to should have a relative or absolute path that includes the subUrl, otherwise the redirect
-	// will send the user to the wrong location
-	if hs.Cfg.AppSubUrl != "" && !strings.HasPrefix(to.Path, hs.Cfg.AppSubUrl) && !strings.HasPrefix(to.Path, "/"+hs.Cfg.AppSubUrl) {
+	if hs.Cfg.AppSubUrl != "" && !strings.HasPrefix(to.Path, hs.Cfg.AppSubUrl+"/") {
```
Add comment explaining logic here.

briangann replied:
looks like comment was removed...

papagian replied:
I have put back the comment

kylebrandt commented on pkg/api/login_oauth.go:
```diff
@@ -217,9 +217,12 @@ func (hs *HTTPServer) OAuthLogin(ctx *m.ReqContext) {
-		ctx.Redirect(redirectTo)
-		return
+		if err := hs.validateRedirectTo(redirectTo); err == nil {
+			middleware.DeleteCookie(ctx.Resp, "redirect_to", hs.cookieOptionsFromCfg)
```
happen to understand this enough to add a comment explaining why the cookie is deleted?

briangann replied:
the cookie is no longer needed since it held the redirect_to location.  this code path is going to send the browser to the new location, and should not retain the redirect_to cookie (could end up in a redirect loop, but mainly it is no longer needed).

papagian replied:
exactly; the cookie is deleted because it's has fulfilled its purpose.

kylebrandt commented:
Don't understand this enough to approve it, but an exemplar pull request message :-)

briangann reviewed (dismissed):
looks great!

consideRatio commented:
I have not fully understood the details here, but I think it makes sense to use the logic from #22265 alongside this PR that allows cookies to be passed along to `/grafana/` and `/grafana` paths and not only the former. My example for why I think so is the following.

- Assume I have a grafana_session cookie already, but without #22265's logic, it's path value remains at `/grafana/` as I understand it, because I don't think this PR influence the `grafana_session` cookie's path value.
- Assume I visit `mydomain.org/grafana`, then my `grafana_session` cookie isn't passed along because I didn't include a trailing `/` in my host header, so I get redirected to /login instead of directly visiting Grafana's root path even though I'm already logged in.

Thanks for working on this!

papagian commented:
> I have not fully understood the details here, but I think it makes sense to use the logic from #22265 alongside this PR that allows cookies to be passed along to `/grafana/` and `/grafana` paths and not only the former. My example for why I think so is the following.
> 
> * Assume I have a grafana_session cookie already, but without #22265's logic, it's path value remains at `/grafana/` as I understand it, because I don't think this PR influence the `grafana_session` cookie's path value.

In that case, [this](https://github.com/grafana/grafana/blob/ab3d7fe56e49f856dd999e1b279ad94f2a5ce35f/pkg/api/login.go#L92) validation returns error and the redirect_to cookie is ignored and delete it and the logged in user is navigated to the home page.

> * Assume I visit `mydomain.org/grafana`, then my `grafana_session` cookie isn't passed along because I didn't include a trailing `/` in my host header, so I get redirected to /login instead of directly visiting Grafana's root path even though I'm already logged in.
> 

This is true.

> Thanks for working on this!

@consideRatio 
Your concerns are fair and I want also to test your modifications on top of these. I'm only a bit reluctant changing the cookie path from what is used to be until today because we had faced a number of issues related to cookie handling and there can be some cases we can't think of right now.

It's also a bit concerning that two different people have difficulty understanding the modifications in this PR therefore I would love some feedback from anybody with better insight than me.

papagian commented:
> > * Assume I visit `mydomain.org/grafana`, then my `grafana_session` cookie isn't passed along because I didn't include a trailing `/` in my host header, so I get redirected to /login instead of directly visiting Grafana's root path even though I'm already logged in.
> 
> This is true.

I have pushed a modification for fixing this.

consideRatio commented:
> > > * Assume I visit `mydomain.org/grafana`, then my `grafana_session` cookie isn't passed along because I didn't include a trailing `/` in my host header, so I get redirected to /login instead of directly visiting Grafana's root path even though I'm already logged in.
> > 
> > 
> > This is true.
> 
> I have pushed a modification for fixing this.

The fix will still make `/grafana?and=query-params` redirect to `/grafana/login`, but from `/grafana/login`, it will realize that the redirect_to cookie specifies that it should go to `/grafana` instead of `/grafana/`, so it will say "no! this is invalid" and then assume the user wants to go to `/grafana/` where the authentication cookie is actually passed along.

It makes a visit to `/grafana` a bit less invalid, but it still disrupts the user by bounces that truncates potential query parameters.

> Your concerns are fair and I want also to test your modifications on top of these. I'm only a bit reluctant changing the cookie path from what is used to be until today because we had faced a number of issues related to cookie handling and there can be some cases we can't think of right now.

This is the key discussion in my mind. Cookie handling is indeed a delicate and complicated matter. I have never really felt confident about them, but after the references you found I've started to feel that I understand them somewhat. That makes me want to tackle the question if the change suggested in #22265 going to cause issues head on. If it isn't going to cause issues, it would actually solve the issue I described above with the workaround in 4e52076 and remove the need to bounce the user to `/home/login` once just to get redirected a place where the cookie is passed.

marefr commented on public/app/core/components/Login/LoginCtrl.tsx:
```diff
@@ -104,9 +104,17 @@ export class LoginCtrl extends PureComponent<Props, State> {
     // Use window.location.href to force page reload
     if (params.redirect && params.redirect[0] === '/') {
-      window.location.href = config.appSubUrl + params.redirect;
+      if (config.appSubUrl !== '' && !params.redirect.startsWith(config.appSubUrl)) {
```
Do we need frontend logic at all for redirect, couldn't the backend make sure to always return the correct url? Wait why are we using redirect query param here - thought we always had redirect in cookie?

marefr reviewed (commented):
Sorry being late to the party. I probably lack a lot of context, but added some comments/questions.

papagian commented on pkg/middleware/auth.go:
```diff
@@ -47,7 +47,11 @@ func notAuthorized(c *m.ReqContext) {
 
-	WriteCookie(c.Resp, "redirect_to", url.QueryEscape(c.Req.RequestURI), 0, newCookieOptions)
+	redirectTo := c.Req.RequestURI
+	if setting.AppSubUrl != "" && !strings.HasPrefix(redirectTo, setting.AppSubUrl) {
```
> Why doesn't c.Req.RequestURI include AppSubUrl?

As far as I understand, the user can request whatever URL.
For instance, the user can request `http://localhost:3000` even though the subpath is `/grafana`.

In addition to this, we always prepended the subpath before [this](https://github.com/grafana/grafana/commit/a7f96ea98e34da8253f7bb071ae6e2bf7435e7b0#diff-c6731ee8c1631796c076df9ae9e4d2c7) change which was introduced for fixing [this](https://github.com/grafana/grafana/pull/21652#issuecomment-582601735) problematic situation.

Probably now that we have validation for the redirects (and if it's working correctly) we no more need to prepend it
because [invalid redirect paths are ignored](4e52076b9dc8c312b2b9323621274a6b0de45601).
But I think it's safer to keep it here, since it appends it only if it's missing.

papagian committed:
Revert "API: Fix redirect issue when configured to use a subpath (#21652)" (#22671)

This reverts commit 0e2d874ecf9277dcc17d562e05271917efc8b595.

papagian committed:
Fix redirect validation (#22675)

* Fix redirect validation

* Fix tests

* Add removed redirect validation comment

* Chore: Add test for parse of app url and app sub url

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

papagian committed:
Fix redirect: prepend subpath only if it's missing (#22676)

* Prepand subpath in redirect URL only if it's missing

* Change frontend to prepend the subpath only if needed

papagian committed:
Validate redirect in login oauth (#22677)

papagian committed:
Fix invalid redirect for authenticated user (#22678)

* Fix login view in case of invalid redirect

If the user is authenticated and
redirect_to cookie points to an invalid path,
the user should be navigated to the home page
instead of rendering the login page with error.

* Fix  TestLoginViewRedirect

marefr committed:
Login: Use correct path for OAuth logos

marefr reviewed (approved):
A: Logout scenario with anonymous auth disabled:
1. Login 
2. Browse dashboard X
3. Clear cookies
4. Refresh browser. Should be redirected to login page
5. Login 
6. Should be redirected to dashboard X

B: Anonymous scenario:
1. Browse dashboard X
2. Click on login in sidemenu and redirect to login page
3. Login 
4. Should be redirected to dashboard X

Root url default without sub path using Grafana auth:
A: Works as expected
B: Works as expected

Root url default without sub path using OAuth:
A: Works as expected
**B: No, redirected to home page**

Root url http://localhost:3000/grafana with serve_from_sub_path using Grafana auth:
A: Works as expected
B: Works as expected

Root url http://localhost:3000/grafana with serve_from_sub_path using OAuth:
A: Works as expected
**B: No, redirected to home page**

Root url http://grafana.myproxy.com:10080/grafana without serve_from_sub_path using nginx and Grafana auth:
A: Works as expected
B: Works as expected

Root url http://grafana.myproxy.com:10080/grafana without serve_from_sub_path using nginx and OAuth:
A: Works as expected
**B: No, redirected to home page**

Also confirmed this resolves #22227

@papagian Not sure if B scenario using OAuth is a new or existing regression? We could create a new issue if there's no existing one and fix that for 6.7 stable maybe?

papagian commented:
> @papagian Not sure if B scenario using OAuth is a new or existing regression? We could create a new issue if there's no existing one and fix that for 6.7 stable maybe?

I was able to reproduce it in 6.5.2 and 6.6.2 and therefore I have filed a new [issue](https://github.com/grafana/grafana/issues/22709)
